float32 x_dist
float32 y_dist
float32 euc_dist